### PR TITLE
Load CascasdeClassifier from cv::FileStorage.

### PR DESCRIPTION
**Pull request by edrosten**
### This pullrequest changes

CascadeClassifier currently allows loading from files using the load method. Internally this opens a FileStorage object and loads from that. This PR exposes that functionality to the public interface. The purpose is to allow the load() to work on all sources supported by FileStorage, for example in memory storage.

### CODE REVIEW

File: modules/objdetect/include/opencv2/objdetect.hpp [hunk 1]
```diff
@@ -241,6 +242,7 @@ class CV_EXPORTS_W CascadeClassifier
     traincascade application.
      */
     CV_WRAP bool load( const String& filename );
+    CV_WRAP bool load( FileStorage& filename );
```
Comment (terfendail) [inline modules/objdetect/include/opencv2/objdetect.hpp]: Could you please remove CV_WRAP here? Precommit build is failed due to inability to provide python bindings for FileStorage object.

File: modules/core/include/opencv2/core/persistence.hpp [hunk 2]
```diff
@@ -1345,6 +1346,19 @@ inline String::String(const FileNode& fn): cstr_(0), len_(0) { read(fn, *this, *
 
 //! @endcond
 
+/** @brief Loads an object from a filestorage.
+
+This performs the same functionality as @cvLoad, but does it directly from a FileStorage object.
+@param filename File storage object
```
Comment (terfendail) [inline modules/core/include/opencv2/core/persistence.hpp]: This parameter should be named `filestorage`. Also could you please remove at sign before`cvLoad` in previous line? It looks like doxygen considered it as internal command.

File: modules/core/include/opencv2/core/persistence.hpp [hunk 3]
```diff
@@ -1345,6 +1346,19 @@ inline String::String(const FileNode& fn): cstr_(0), len_(0) { read(fn, *this, *
...
+@param memstorage Memory storage for dynamic structures, such as CvSeq or CvGraph . It is not used
+for matrices or images.
+@param name Optional object name. If it is NULL, the first top-level object in the storage will be
+loaded.
+@param real_name Optional output parameter that will contain the name of the loaded object
+(useful if name=NULL )
+ */
+CV_EXPORTS void* cvLoadFileStorage(FileStorage& filestorage, CvMemStorage* memstorage, const char* name, const char** real_name);
```
Comment (alalek) [inline modules/core/include/opencv2/core/persistence.hpp]: Adding of new public entries for legacy C-like API must be avoided.
CvMemStorage and other "C API" things would be removed in OpenCV 4.0, so nobody would write proper tests for this.

File: modules/objdetect/src/cascadedetect.cpp [hunk 4]
```diff
@@ -904,21 +904,24 @@ bool CascadeClassifierImpl::empty() const
...
     if( !fs.isOpened() )
         return false;
 
     if( read_(fs.getFirstTopLevelNode()) )
         return true;
 
-    fs.release();
-
-    oldCascade.reset((CvHaarClassifierCascade*)cvLoad(filename.c_str(), 0, 0, 0));
+    oldCascade.reset((CvHaarClassifierCascade*)cvLoadFileStorage(fs, 0, 0, 0));
```
Comment (alalek) [inline modules/objdetect/src/cascadedetect.cpp]: Do we really need to support loading of old cascades by this way?

File: modules/objdetect/include/opencv2/objdetect.hpp [hunk 5]
```diff
@@ -177,6 +177,7 @@ class CV_EXPORTS_W BaseCascadeClassifier : public Algorithm
     virtual ~BaseCascadeClassifier();
     virtual bool empty() const CV_OVERRIDE = 0;
     virtual bool load( const String& filename ) = 0;
+    virtual bool load( FileStorage& filestorage) = 0;
```
Comment (alalek) [inline modules/objdetect/include/opencv2/objdetect.hpp]: New functionality should be covered by tests.

**Comment by vpisarev** [comment]
@edrosten, thank you for the contribution! As @alalek said, we are going to remove C-API in OpenCV 4.0 and the persistence functionality (XML/YAML I/O) will be rewritten using C++ structures. I suggest submit the revised PR after it's done

**Pull request closed without merging**